src/openrc/rc.c

```diff
@@ -349,22 +356,44 @@ static char *get_krunlevel(void)
 static void
 add_pid(pid_t pid)
 {
+	sigset_t sm, oldsm;
+
 	RC_PID *p = xmalloc(sizeof(*p));
 	p->pid = pid;
+
+	/* this list might be accessed inside the signal handler. so block
+	 * singnals to ensure it doesn't get into an inconsistent state.
+	 */
+	sigfillset(&sm);
+	sigprocmask(SIG_SETMASK, &sm, &oldsm);
 	LIST_INSERT_HEAD(&service_pids, p, entries);
+	sigprocmask(SIG_SETMASK, &oldsm, NULL);
 }
 
 static void
-remove_pid(pid_t pid)
+remove_pid(pid_t pid, bool inside_signal)
 {
-	RC_PID *p;
-
-	LIST_FOREACH(p, &service_pids, entries)
-	    if (p->pid == pid) {
-		    LIST_REMOVE(p, entries);
-		    free(p);
-		    return;
-	    }
+	sigset_t sm, oldsm;
+	RC_PID *p, *tmp;
+
+	sigfillset(&sm);
+	LIST_FOREACH(p, &service_pids, entries) {
+		if (p->pid == pid) {
+			/* same rationale for blocking signal as add_pid() */
+			sigprocmask(SIG_SETMASK, &sm, &oldsm);
```

Comment:
LIST_FOREACH itself isn't safe if a link is removed halfway through, this needs to be blocked through the whole loop

Reply:
Hmm, good catch. I'll update it on the next iteration.